src/DataStructures/Tensor/Expressions/Product.hpp

```diff
@@ -56,14 +61,16 @@ template <typename T1, typename T2, template <typename...> class IndexList1,
 struct OuterProduct<T1, T2, IndexList1<Indices1...>, IndexList2<Indices2...>,
                     ArgsList1<Args1...>, ArgsList2<Args2...>>
     : public TensorExpression<
-          OuterProduct<T1, T2>, typename T1::type,
+          OuterProduct<T1, T2>, typename detail::OuterProductType<T1, T2>::type,
           typename detail::OuterProductType<T1, T2>::symmetry,
           typename detail::OuterProductType<T1, T2>::index_list,
           typename detail::OuterProductType<T1, T2>::tensorindex_list> {
-  static_assert(std::is_same<typename T1::type, typename T2::type>::value,
+  static_assert(std::is_same<typename T1::type, typename T2::type>::value or
```

Comment:
This should mean we support:
`double * TE<double, indices....>`
`double * TE<DataVector, indices....>`
Meaning, `double * (any derived TE, not just TensorAsExpression, whose type is double or DataVector)`

But is this also supposed to be supported?:
`TensorAsExpression<double> * TE<DataVector, indices...>`
Meaning, `(rank 0 Tensor storing a double) * (any derived TE, not just TensorAsExpression, whose type is double or DataVector)`

Reply:
No, I think only `double * TE` should be supported. I don't really see how `R() * TE<DataVector>` with R being a double is useful. I'd expect the double to be some constant, e.g. 0.5 or 0.25, or we are multiplying by an entire field.  The more general case just means relaxing a static assert, right? If that's the case I'd say implement the restricted case, we can generalize easily later if necessary

Reply:
Sounds good!